### PR TITLE
ds - updated rr controller and tests

### LINKED ISSUE #28: Implement GET (show), PUT (edit), DELETE, and POST (create) endpoints (see above for endpoint suggestions)

### LINKED ISSUE #30: All the endpoints are implemented and tested in `RecommendationRequestController.java` and `RecommendationRequestControllerTests.java` and tested on Swagger and dokku-qa

### PR DESCRIPTION
**Pull request by divyagsubramonian**
### Description:
- There are endpoints for each of the CRUD actions necessary to create, read, update, and delete a recommendation request that have been tested on swagger
  - GET api/recommendationrequest/{id}
  - PUT api/recommendationrequest/{id}
  - DELETE api/recommendationrequest/{id}
  - POST api/recommendationrequest/post
- GET by professorName api/recommendationrequest/{professorName} **and** GET by studentName api/recommendationrequest/{studentName} have _not_ been implemented in this PR

### Tests:
 - Passes all tests

### Dokku:
 - Deployed at https://ds-rr-controller.dokku-07.cs.ucsb.edu/
 - Log in, then use swagger to test the backend

### Related Issues:
 - Dependent on #24 and #25 
 - Closes #28 
 - Closes #30

### CODE REVIEW

**Review by lodetrick — APPROVED** [review]
LGTM! I can see all the fields, and the tests look good. We might want to create an issue later to get rid of the name fields and replace those with user api calls, but other than that it looks good!

File: src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+     * 
+     * @return an iterable of RecommendationRequest
+     */
+    @Operation(summary= "List all recommendation requests")
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @GetMapping("/all")
+    public Iterable<RecommendationRequest> allRecommendationRequests() {
+        Iterable<RecommendationRequest> requests = recommendationRequestRepository.findAll();
+        return requests;
+    }
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java:49]: No... if this endpoint is for ROLE_USER it should look up the id of the current user and the return ONLY the recommendation requests for THIS user.  Not a findAll but a findByUserId(currentUser.getId())

File: src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+     * @param professorName
+     * @param requestType      
+     * @param details           
+     * @param submissionDate
+     * @param completionDate
+     * @param status
+     * @return                  
+     */
+    @Operation(summary= "Create a new request")
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java:66]: No.  Requests are created by users, not admins.

Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java:66]:
```suggestion
    @PreAuthorize("hasRole('ROLE_USER')")
```

File: src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    @PostMapping("/post")
+    public RecommendationRequest postRecommendationRequest(
+            @Parameter(name="requesterEmail") @RequestParam String requesterEmail,
+            @Parameter(name="requesterName") @RequestParam String requesterName,
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java:70]: No. You should get the currentUser.getId() and then set the requesterId inside the controller method.(there should NOT be a database field for requesterEmail or requesterName, only requestId).

File: src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+            @Parameter(name="professorEmail") @RequestParam String professorEmail,
+            @Parameter(name="professorName") @RequestParam String professorName,
+            @Parameter(name="requestType") @RequestParam String requestType,
+            @Parameter(name="details") @RequestParam String details,
+            @Parameter(name="submissionDate", description="date (in iso format, e.g. YYYY-mm-ddTHH:MM:SS; see https://en.wikipedia.org/wiki/ISO_8601)") @RequestParam("submissionDate") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime submissionDate,
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java:75]: No.  The submission date should be set inside the controller from a call to get the current date and time, or better yet, using the methods described here:

https://ucsb-cs156.github.io/topics/spring_boot/spring_boot_database.html#timestamps-on-database-rows

File: src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+     * @return                  
+     */
+    @Operation(summary= "Create a new request")
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @PostMapping("/post")
+    public RecommendationRequest postRecommendationRequest(
+            @Parameter(name="requesterEmail") @RequestParam String requesterEmail,
+            @Parameter(name="requesterName") @RequestParam String requesterName,
+            @Parameter(name="professorEmail") @RequestParam String professorEmail,
+            @Parameter(name="professorName") @RequestParam String professorName,
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java:72]: No.  The userId of the professsor should be passed in via a dropdown menu that shows all users in the users database that have the professor role.  From that, we can get the email and name of the professor.

File: src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+            @Parameter(name="submissionDate", description="date (in iso format, e.g. YYYY-mm-ddTHH:MM:SS; see https://en.wikipedia.org/wiki/ISO_8601)") @RequestParam("submissionDate") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime submissionDate,
+            @Parameter(name="completionDate", description="date (in iso format, e.g. YYYY-mm-ddTHH:MM:SS; see https://en.wikipedia.org/wiki/ISO_8601)") @RequestParam("completionDate") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime completionDate,
+            @Parameter(name="status") @RequestParam String status)
+            throws JsonProcessingException {
+
+        // For an explanation of @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
+        // See: https://www.baeldung.com/spring-date-parameters
+
+        // log.info("dateRequested={}", dateRequested);
+        // log.info("dateNeeded={}", dateNeeded);
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java:84]: Why are these commented out lines still in a PR that has been marked as Ready-For-Staff?

https://ucsb-cs156.github.io/topics/code_reviews/commented_out_code.html#why-should-commented-out-code-be-removed

File: src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+        RecommendationRequest recommendationRequest = new RecommendationRequest();
+        recommendationRequest.setRequesterEmail(requesterEmail);
+        recommendationRequest.setRequesterName(requesterName);
+        recommendationRequest.setProfessorEmail(professorEmail);
+        recommendationRequest.setProfessorName(professorName);
+        recommendationRequest.setRequestType(requestType);
+        recommendationRequest.setDetails(details);
+        recommendationRequest.setSubmissionDate(submissionDate);
+        recommendationRequest.setCompletionDate(completionDate);
+        recommendationRequest.setStatus(status);
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java:95]: Status should be hard coded as `pending` on a newly created request.  The end user doesn't get to "choose" what status a new request is in.  It is *always* in the pending state, and only the professor gets to change it to either `accepted` or `denied`

File: src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+        RecommendationRequest recommendationRequest = new RecommendationRequest();
+        recommendationRequest.setRequesterEmail(requesterEmail);
+        recommendationRequest.setRequesterName(requesterName);
+        recommendationRequest.setProfessorEmail(professorEmail);
+        recommendationRequest.setProfessorName(professorName);
+        recommendationRequest.setRequestType(requestType);
+        recommendationRequest.setDetails(details);
+        recommendationRequest.setSubmissionDate(submissionDate);
+        recommendationRequest.setCompletionDate(completionDate);
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java:94]: This should be left as null until the PROFESSOR marks it as complete, and then it should be set to the date on which that happens.

File: src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+        // log.info("dateNeeded={}", dateNeeded);
+
+        RecommendationRequest recommendationRequest = new RecommendationRequest();
+        recommendationRequest.setRequesterEmail(requesterEmail);
+        recommendationRequest.setRequesterName(requesterName);
+        recommendationRequest.setProfessorEmail(professorEmail);
+        recommendationRequest.setProfessorName(professorName);
+        recommendationRequest.setRequestType(requestType);
+        recommendationRequest.setDetails(details);
+        recommendationRequest.setSubmissionDate(submissionDate);
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java:93]: This should be set to current date/time or else set using the @CreatedDate annotation

File: src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+        // See: https://www.baeldung.com/spring-date-parameters
+
+        // log.info("dateRequested={}", dateRequested);
+        // log.info("dateNeeded={}", dateNeeded);
+
+        RecommendationRequest recommendationRequest = new RecommendationRequest();
+        recommendationRequest.setRequesterEmail(requesterEmail);
+        recommendationRequest.setRequesterName(requesterName);
+        recommendationRequest.setProfessorEmail(professorEmail);
+        recommendationRequest.setProfessorName(professorName);
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java:90]: Why is there still a professorEmail and professorName field? This should be replaced with professorId

File: src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+
+        // For an explanation of @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
+        // See: https://www.baeldung.com/spring-date-parameters
+
+        // log.info("dateRequested={}", dateRequested);
+        // log.info("dateNeeded={}", dateNeeded);
+
+        RecommendationRequest recommendationRequest = new RecommendationRequest();
+        recommendationRequest.setRequesterEmail(requesterEmail);
+        recommendationRequest.setRequesterName(requesterName);
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java:88]: Why is there still a requesterEmail and requesterName field?  This should be replaced with requesterId.

File: src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+    @Operation(summary= "Get a single request")
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @GetMapping("")
+    public RecommendationRequest getById(
+            @Parameter(name="id") @RequestParam Long id) {
+        RecommendationRequest recommendationRequest = recommendationRequestRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException(RecommendationRequest.class, id));
+
+        return recommendationRequest;
+    }
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java:117]: You *also* need to see if this recommendation request belongs to this user!  They should NOT be able to see requests that are for other students.

File: src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @DeleteMapping("")
+    public Object deleteRecommendationRequest(
+            @Parameter(name="id") @RequestParam Long id) {
+        RecommendationRequest recommendationRequest = recommendationRequestRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException(RecommendationRequest.class, id));
+
+                recommendationRequestRepository.delete(recommendationRequest);
+        return genericMessage("RecommendationRequest with id %s deleted".formatted(id));
+    }
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java:135]: Users (not just admins) should be able to delete their OWN (but only their own) requests.  Please adjust the authorization, and add a check not just that the id is in the database, but also that the requesterId matches the currentUser.getId().

File: src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+
+    /**
+     * Update a single request
+     * 
+     * @param id       id of the request to update
+     * @param incoming the new date
+     * @return the updated date object
+     */
+    @Operation(summary= "Update a single request")
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/rec/controllers/RecommendationRequestController.java:145]: Users should be able to update their own request, and admins actually should NOT be able to.

But only their own requests, and only the details field, NOTHING ELSE.

**Review by pconrad — CHANGES_REQUESTED** [review]
It does not appear that the author has taken the time to think through the  user stories involved here.  It is *so important* to keep the end user in mind, and what their *goal* is.

It is also important to really think carefully about the user *should* be able to do, and what they *should not* be able to do.

Read through the code, and the comments, and go BACK to the original issues (not the ones written up by the team, which contain many misunderstandings.)

**Pull request closed without merging**